Add initial SQL support for non-expression sketch postaggs

jon-wei opened the pull request:
This PR adds support for using PostAggregators on Datasketches aggregators in SQL, by adding new SQL operators that accept quantile, HLL, and theta sketches.

To support this, the SQL system has been updated to allow operators to be converted into non-expression-based PostAggregators. This is accomplished by adding a new `PostAggregatorVisitor` class and adding new variants of the existing RexNode conversion methods that accept a post agg visitor.

This PR does not add support for all current sketch postaggregators (such as `DoublesSketchToCDFPostAggregator`) and the tuple sketch is currently not supported, I plan to  handle this in a follow on PR. 

This PR has:
- [x] been self-reviewed.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [x] added unit tests or modified existing tests to cover new code paths.

jon-wei commented:
Marking WIP temporarily, need to update the docs

jon-wei commented:
Updated docs, removing WIP tag

ccaominh commented on docs/querying/sql.md:
```diff
@@ -363,6 +366,45 @@ All 'array' references in the multi-value string function documentation can refe
+|--------|-----|
+|`THETA_SKETCH_ESTIMATE(expr)`|Returns the distinct count estimate from a theta sketch. `expr` must return a theta sketch.|
+|`THETA_SKETCH_ESTIMATE_WITH_ERROR_BOUNDS(expr, errorBoundsStdDev)`|Returns the distinct count estimate and error bounds from a theta sketch. `expr` must return a theta sketch.|
+|`THETA_SKETCH_UNION([size], expr0, expr1, ...)`|Returns a union of theta sketches, where each input expression must return an theta sketch. The `size` can be optionally specified as the first parameter.|
```
Typo: an theta sketch -> a theta sketch

(same for following 2 lines)

FYI, it's possible that the spellcheck in travis will have some false positives on the function names. You can add suppress false positives for this file by adding entries after https://github.com/apache/incubator-druid/blob/master/website/.spelling#L1307

jon-wei replied:
thanks, fixed, looks like there were no spellcheck errors

jihoonson commented:
Added `Design Review` since this PR adds new SQL functions.

jihoonson commented on extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/hll/sql/HllSketchSetUnionOperatorConversion.java:
```diff
@@ -0,0 +1,137 @@
+      if (convertedPostAgg == null) {
+        if (operandCounter == 0) {
+          try {
+            lgK = RexLiteral.intValue(operand);
```
Should we check the kind of `operand` is `LITERAL` here before we call `intValue()`? Otherwise, this line can throw an `AssertionError` which is not caught by the below `catch`.

jon-wei replied:
Added sqlkind checks here and in similar places

jihoonson commented on extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/hll/sql/HllSketchSetUnionOperatorConversion.java:
```diff
@@ -0,0 +1,137 @@
+          try {
+            lgK = RexLiteral.intValue(operand);
+          }
+          catch (RuntimeException re) {
```
It should be better to narrow the exception here, so that we won't catch some exceptions unintentionally. Looks like `ClassCastException` is the only possible `RuntimeException`.

jon-wei replied:
Narrowed this to `ClassCastException` here and elsewhere

jihoonson commented on extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/theta/sql/ThetaSketchSetBaseOperatorConversion.java:
```diff
@@ -0,0 +1,140 @@
+      if (convertedPostAgg == null) {
+        if (operandCounter == 0) {
+          try {
+            size = RexLiteral.intValue(operand);
```
Same here for literal check and exception handling.

jon-wei replied:
Fixed

jihoonson commented on extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/theta/sql/ThetaSketchSetBaseOperatorConversion.java:
```diff
@@ -0,0 +1,140 @@
+
+  public String getSetOperationName()
+  {
+    throw new UnsupportedOperationException("getSetOperationName() is not implemented.");
```
Suggest `public abstract String getSetOperationName()`.

jon-wei replied:
Changed this to abstract method

jon-wei commented on docs/querying/sql.md:
```diff
@@ -363,6 +366,44 @@ All 'array' references in the multi-value string function documentation can refe
+
+|Function|Notes|
+|--------|-----|
+|`HLL_SKETCH_ESTIMATE(expr)`|Returns the distinct count estimate from an HLL sketch. `expr` must return an HLL sketch.|
```
Hmm, i actually like the new functions added in this PR more than `APPROX_COUNT_DISTINCT_DS_HLL`, I think they express a clearer boundary between the aggregated sketch object and the operations that can be performed on said object. The new functions names also have a 1-to-1 correspondence with the underlying native aggs/postaggs which I think is nice.

If not for backwards compatibility, I would actually consider dropping `APPROX_COUNT_DISTINCT_DS_HLL`, it was introduced as a way to allow people to use the datasketches implementations before we supported postaggs in SQL (with this patch).

jihoonson replied:
After my second look, the new ones do look better to me too. Thanks.

jon-wei commented on docs/querying/sql.md:
```diff
@@ -363,6 +366,44 @@ All 'array' references in the multi-value string function documentation can refe
+|Function|Notes|
+|--------|-----|
+|`HLL_SKETCH_ESTIMATE(expr)`|Returns the distinct count estimate from an HLL sketch. `expr` must return an HLL sketch.|
+|`HLL_SKETCH_ESTIMATE_WITH_ERROR_BOUNDS(expr, [numStdDev])`|Returns the distinct count estimate and error bounds from an HLL sketch. `expr` must return an HLL sketch. An optional `numStdDev` argument can be provided.|
```
Within the context of this sketch and its documentation (https://druid.apache.org/docs/latest/development/extensions-core/datasketches-hll.html), I think the meaning of `estimate` is clear

jihoonson replied:
Ok, thanks.

jon-wei commented on docs/querying/sql.md:
```diff
@@ -363,6 +366,44 @@ All 'array' references in the multi-value string function documentation can refe
+
+|Function|Notes|
+|--------|-----|
+|`THETA_SKETCH_ESTIMATE(expr)`|Returns the distinct count estimate from a theta sketch. `expr` must return a theta sketch.|
```
My reasoning here is the same as in this comment: https://github.com/apache/incubator-druid/pull/8487/files#r335716107

jon-wei commented on docs/querying/sql.md:
```diff
@@ -363,6 +366,44 @@ All 'array' references in the multi-value string function documentation can refe
+|Function|Notes|
+|--------|-----|
+|`THETA_SKETCH_ESTIMATE(expr)`|Returns the distinct count estimate from a theta sketch. `expr` must return a theta sketch.|
+|`THETA_SKETCH_ESTIMATE_WITH_ERROR_BOUNDS(expr, errorBoundsStdDev)`|Returns the distinct count estimate and error bounds from a theta sketch. `expr` must return a theta sketch.|
```
I think given the context and other docs (https://druid.apache.org/docs/latest/development/extensions-core/datasketches-theta.html), `estimate` is clear here

jon-wei commented on extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/hll/HllSketchToEstimatePostAggregator.java:
```diff
@@ -0,0 +1,135 @@
+  public byte[] getCacheKey()
+  {
+    return new CacheKeyBuilder(PostAggregatorIds.HLL_SKETCH_TO_ESTIMATE_CACHE_TYPE_ID)
+        .appendString(name)
```
Removed, thanks

jon-wei commented on extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/hll/sql/HllSketchEstimateOperatorConversion.java:
```diff
@@ -0,0 +1,97 @@
+
+public class HllSketchEstimateOperatorConversion extends DirectOperatorConversion
+{
+  private static String FUNCTION_NAME = "HLL_SKETCH_ESTIMATE";
```
Added final here and in similar places

jon-wei commented on extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/hll/sql/HllSketchApproxCountDistinctSqlAggregator.java:
```diff
@@ -0,0 +1,211 @@
+      boolean finalizeAggregations
+  )
+  {
+    // Don't use Aggregations.getArgumentsForSimpleAggregator, since it won't let us use direct column access
```
Refactored this to use a base abstract class (also did the same for theta sketches).

I also renamed the aggregators that produce a sketch object to `*SketchObjectSqlAggregator` from `*SketchSqlAggregator`

jon-wei commented on extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/hll/sql/HllSketchSetUnionOperatorConversion.java:
```diff
@@ -0,0 +1,137 @@
+          factory -> Calcites.createSqlType(factory, SqlTypeName.OTHER)
+      ),
+      null,
+      OperandTypes.VARIADIC,
```
Changed to suggested

jon-wei commented on extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/theta/sql/ThetaSketchSetBaseOperatorConversion.java:
```diff
@@ -0,0 +1,140 @@
+            factory -> Calcites.createSqlType(factory, SqlTypeName.OTHER)
+        ),
+        null,
+        OperandTypes.VARIADIC,
```
Changed to suggested

jon-wei commented on extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/hll/sql/HllSketchSetUnionOperatorConversion.java:
```diff
@@ -0,0 +1,144 @@
+      SqlFunctionCategory.USER_DEFINED_FUNCTION
+  );
+
+  public HllSketchSetUnionOperatorConversion()
```
Removed

jon-wei commented on sql/src/main/java/org/apache/druid/sql/calcite/expression/PostAggregatorVisitor.java:
```diff
@@ -0,0 +1,70 @@
+    return oldVal;
+  }
+
+  public int getCounter()
```
Removed

jon-wei commented on sql/src/main/java/org/apache/druid/sql/calcite/expression/PostAggregatorVisitor.java:
```diff
@@ -0,0 +1,70 @@
+  {
+    int oldVal = counter;
+    counter++;
+    return oldVal;
```
Changed to suggested

jon-wei commented:
@jihoonson Addressed latest comments, also added a "round" parameter to the HLL sketch estimate postagg, and fixed the `numStdDev` param for `HLL_SKETCH_ESTIMATE_WITH_ERROR_BOUNDS` to actually be optional (it's optional in the docs and the underlying postagg).
